File: Appointment/AppointmentDetailTableViewCell.m
//
//  AppointmentDetailTableViewCell.m
//  Appointment
//
//  Created by Jeffrey Moran on 6/8/17.
//  Copyright © 2017 Jeff Moran. All rights reserved.
//

#import "AppointmentDetailTableViewCell.h"

@implementation AppointmentDetailTableViewCell

- (void)awakeFromNib {
	[super awakeFromNib];
	// Initialization code
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
	self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
	
	if (self) {
		[self addSubviews];
		[self setUpConstraints];
	}
	
	return self;
}

- (void)setAppointment:(Appointment *)anAppointment with:(NSIndexPath *)indexPath {
	[self setUpAppointmentValues:anAppointment with:indexPath];
}

- (void)addSubviews {
	appointmentHeaderLabel = [[UILabel alloc] init];
	appointmentHeaderLabel.translatesAutoresizingMaskIntoConstraints = NO;
	appointmentHeaderLabel.numberOfLines = 2;
	appointmentHeaderLabel.adjustsFontSizeToFitWidth = YES;
	appointmentHeaderLabel.textAlignment = NSTextAlignmentRight;
	appointmentHeaderLabel.textColor = [UIColor darkGrayColor];
	appointmentHeaderLabel.font = [UIFont systemFontOfSize:13];
	
	appointmentValueLabel = [[UILabel alloc] init];
	appointmentValueLabel.translatesAutoresizingMaskIntoConstraints = NO;
	appointmentValueLabel.font = [UIFont boldSystemFontOfSize:18];
	appointmentValueLabel.numberOfLines = 0;
	
	[self.contentView addSubview:appointmentHeaderLabel];
	[self.contentView addSubview:appointmentValueLabel];
}

- (void)setUpConstraints {
	[NSLayoutConstraint
	 activateConstraints:@[
						   [self.contentView.heightAnchor constraintGreaterThanOrEqualToConstant:50],
						   
						   [appointmentHeaderLabel.leftAnchor constraintEqualToAnchor:self.contentView.leftAnchor constant: 5],
						   [appointmentHeaderLabel.topAnchor constraintEqualToAnchor:self.contentView.topAnchor],
						   [appointmentHeaderLabel.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor],
						   [appointmentHeaderLabel.widthAnchor constraintEqualToConstant:80],
						   
						   [appointmentValueLabel.leftAnchor constraintEqualToAnchor:appointmentHeaderLabel.rightAnchor constant: 5],
						   [appointmentValueLabel.topAnchor constraintEqualToAnchor:self.contentView.topAnchor],
						   [appointmentValueLabel.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor constant: -5],
						   [appointmentValueLabel.rightAnchor constraintEqualToAnchor:self.contentView.rightAnchor constant: -5]
						   ]];
}

- (void)setUpAppointmentValues:(Appointment *)anAppointment with:(NSIndexPath *)indexPath {
	appointmentHeaderLabel.text = anAppointment.appointmentPropertiesHeader[indexPath.row];
	appointmentValueLabel.text = anAppointment.appointmentProperties[indexPath.row];
}

@end

File: Appointment/AppointmentDetailTableViewCell.swift
//
//  AppointmentDetailTableViewCell.swift
//  Appointment
//
//  Created by Jeffrey Moran on 10/15/20.
//  Copyright © 2020 Jeff Moran. All rights reserved.
//

import UIKit

class AppointmentDetailTableViewCell: UITableViewCell {

    // MARK: - Static Constants

    static let reuseIdentifier = "appointmentDetailTableViewCellIdentifier"

    // MARK: - Private Properties

    private lazy var stackView: UIStackView = {
        let stackView = UIStackView()
        stackView.translatesAutoresizingMaskIntoConstraints = false
        stackView.spacing = UIStackView.spacingUseSystem

        return stackView
    }()

    private lazy var appointmentHeaderLabel: UILabel = {
        let label = UILabel()
        label.textColor = .secondaryLabel
        label.textAlignment = .right
        label.numberOfLines = 2
        label.font = .systemFont(ofSize: 14.0)

        return label
    }()

    private lazy var appointmentValueLabel: UILabel = {
        let label = UILabel()
        label.textColor = .label
        label.font = .systemFont(ofSize: 18.0, weight: .semibold)

        return label
    }()

    // MARK: - Initializers

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)

        addSubviews()
        setUpConstraints()
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - Private Methods

    private func addSubviews() {
        contentView.addSubview(stackView)
        stackView.addArrangedSubview(appointmentHeaderLabel)
        stackView.addArrangedSubview(appointmentValueLabel)
    }

    private func setUpConstraints() {
        NSLayoutConstraint.activate([
            appointmentHeaderLabel.widthAnchor.constraint(equalToConstant: 80),

            stackView.leadingAnchor.constraint(equalTo: contentView.readableContentGuide.leadingAnchor),
            stackView.trailingAnchor.constraint(equalTo: contentView.readableContentGuide.trailingAnchor),
            stackView.topAnchor.constraint(equalTo: contentView.readableContentGuide.topAnchor),
            stackView.bottomAnchor.constraint(equalTo: contentView.readableContentGuide.bottomAnchor)

        ])
    }

    // MARK: - Public Methods

    func style(with appointment: Appointment, index: Int) {
        let appointmentDetail = AppointmentDetail(rawValue: index)

        appointmentHeaderLabel.text = appointmentDetail?.headerValue
        appointmentValueLabel.text = appointmentDetail?.detailValue(for: appointment)
    }
}
